src/Sparrow.Server/Allocators.Pool.cs

```diff
@@ -0,0 +1,312 @@
+
+                    // Pointer was holding the marker for the next released block instead. 
+                    allocator._freed[index] = *((BlockPointer*)section.Address);
+                    allocator.InUse += size;
+                    allocator.TotalAllocated += section.Size;
+
+                    return new BlockPointer(section.Address, section.BlockSize, size);
+                }
+            }        
+
+            var ptr = _internalAllocator.Allocate(size);
+            allocator.InUse += size;
+            allocator.Allocated += ptr.Size;
+            allocator.TotalAllocated += ptr.Size;
+
+            return new BlockPointer(ptr.Address, size, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(ref PoolAllocator<TOptions> allocator, ref BlockPointer ptr)
+        {
+            // When MaxPoolSizeInBytes is zero, we are effectively disabling the pooling code generation.
+            // It is useful for cases where composition is done through type chaining. 
+            if (allocator.InUse > allocator._options.MaxPoolSizeInBytes || Allocator.LowMemoryFlag.IsRaised() || allocator._options.MaxPoolSizeInBytes == 0)
+                goto UnlikelyRelease;
+
+            int originalSize = ptr.Size;
+
+            int size = ptr.BlockSize;
+            if (!Bits.IsPowerOfTwo(size))
+                size = Bits.PowerOf2(size) >> 1;
+            
+            var index = Bits.MostSignificantBit(size) - 2; // We use -2 because we are not starting at 0. 
+
+            // Retaining chunks bigger than the max chunk size could clutter the allocator, so we reroute it to the backing allocator.
+            if (index < 0 || index >= allocator._freed.Length)
+                goto UnlikelyRelease;
+
+            var section = allocator._freed[index];
+            if (section.IsValid)
+            {
+                // Copy the section pointer that is already freed to the current memory. 
+                *(BlockPointer*)ptr.Address = section;
+            }
+            else
+            {
+                // Put a copy of the currently released memory block on the front. 
+                *(BlockPointer*)ptr.Address = new BlockPointer();
+            }
+
+            allocator._freed[index] = ptr;
+            allocator.InUse -= originalSize;
+
+            ptr = new Pointer(); // Nullify the pointer
+
+            return;
+
+        UnlikelyRelease:
+            // This should be an unlikely case, when you are running out of memory or over allocated,
+            // all performance guarantees go down the drain. So we dont care if we hit expensive calls
+            // that allows us to release some steam at the cost of hitting cold code. 
+            // https://github.com/dotnet/coreclr/issues/6024
+
+            allocator.InUse -= ptr.Size;
+            allocator.Allocated -= ptr.BlockSize;
+
+            Pointer nakedPtr = new Pointer(ptr.Address, ptr.BlockSize);
```

Comment:
Why do the copy?

Reply:
Because when running in validation mode, that pointer may actually be 'tracked` by a different allocator. For the purpose of the `AllocatorPool` the pointer is 'new` no allocator tracking is needed.